Add mlx docs (#1207)
huggingface/hub-docs: docs/hub/mlx.md

# Using MLX at Hugging Face

> [MLX]([https://github.com/MaartenGr/BERTopic](https://github.com/ml-explore/mlx)) is a model training and serving framework on Apple Silicon made by Apple Research.

Reviewer: ```suggestion
[MLX](https://github.com/ml-explore/mlx) is a model training and serving framework on Apple Silicon made by Apple Research.
```

> [MLX]([https://github.com/MaartenGr/BERTopic](https://github.com/ml-explore/mlx)) is a model training and serving framework on Apple Silicon made by Apple Research.

Reviewer 2: ```suggestion
[MLX]([https://github.com/MaartenGr/BERTopic](https://github.com/ml-explore/mlx)) is a model training and serving framework on Apple Silicon made by Apple.
```

> [MLX]([https://github.com/MaartenGr/BERTopic](https://github.com/ml-explore/mlx)) is a model training and serving framework on Apple Silicon made by Apple Research.

Reviewer 3: ```suggestion
[MLX]([https://github.com/MaartenGr/BERTopic](https://github.com/ml-explore/mlx)) is a model training and serving framework for Apple silicon made by Apple machine learning research.
```

> It comes with variety of examples:

Reviewer 4: ```suggestion
It comes with a variety of examples:
```

> - [Transformer language model](https://github.com/ml-explore/mlx-examples/tree/main/transformer_lm) training.

Reviewer: I'd add reference to these two first:

1. https://github.com/ml-explore/mlx-examples/tree/main/llms/mlx_lm
2. https://github.com/ml-explore/mlx-examples/tree/main/llms/gguf_llm

> - Large-scale text generation with [LLaMA](https://github.com/ml-explore/mlx-examples/tree/main/llms/llama) and
> - finetuning with [LoRA](https://github.com/ml-explore/mlx-examples/tree/main/lora).

Reviewer 4: ```suggestion
- Large-scale text generation with [LLaMA](https://github.com/ml-explore/mlx-examples/tree/main/llms/llama).
- Fine-tuning with [LoRA](https://github.com/ml-explore/mlx-examples/tree/main/lora).
```

- Generating images with [Stable Diffusion](https://github.com/ml-explore/mlx-examples/tree/main/stable_diffusion).
- Speech recognition with [OpenAI's Whisper](https://github.com/ml-explore/mlx-examples/tree/main/whisper).

## Exploring MLX on the Hub

You can find MLX models by filtering at the left of the [models page](https://huggingface.co/models?library=mlx&sort=trending).

> There's also [MLX community](https://huggingface.co/mlx-community) converting and publishing weights for MLX format.

Reviewer 2: ```suggestion
There's also an open [MLX community](https://huggingface.co/mlx-community) of contributors converting and publishing weights for MLX format.
```

> Thanks to MLX Hugging Face Hub integration, you can load MLX models with a few lines of code. 

Reviewer 4: Maybe link to `Using Existing Models` here?

Author: Added at the end

## Installation

MLX comes as a standalone package, and there's a subpackage called MLX-LM with Hugging Face integration for Large Language Models.
To install MLX-LM, you can use the following one-line install through `pip`:

```bash
pip install mlx-lm
```

You can get more information about it [here](https://github.com/ml-explore/mlx-examples/blob/main/llms/README.md#generate-text-with-llms-and-mlx). 

You can install MLX itself as follows.

With `pip`:

```bash
pip install mlx
```

With `conda`:

```bash
conda install -c conda-forge mlx

> ```

Reviewer 4: I'd maybe move this to a later section, otherwise it looks like it's required for HF use in addition to `mlx-lm`.

Author: Can you elaborate? This is sort of what I understood from MLX-LM itself so

Author: Addressed this.

## Using Existing Models

MLX-LM has useful utilities to generate text. The following line directly downloads and loads the model and starts generating text.

```bash

> python -m mlx_lm.generate --model mistralai/Mistral-7B-v0.1 --prompt "hello"

Reviewer 3: We are planning to switch the default model here to the instruct as this often gives bad/unsafe outputs:

```
python -m mlx_lm.generate --model mistralai/Mistral-7B-Instruct-v0.2 --prompt "hello"
```

```

For a full list of generation options, run

```bash
python -m mlx_lm.generate --help
```

You can also load a model and start generating text through Python like below:

```python
from mlx_lm import load, generate

> model, tokenizer = load("mistralai/Mistral-7B-v0.1")

Reviewer 3: Same comment here regarding model:
```suggestion
model, tokenizer = load("mistralai/Mistral-7B-Instruct-v0.2")
```

response = generate(model, tokenizer, prompt="hello", verbose=True)
```

MLX models that are not supported by MLX-LM can easily be downloaded as follows:

```py
pip install huggingface_hub hf_transfer

export HF_HUB_ENABLE_HF_TRANSFER=1
huggingface-cli download --local-dir <LOCAL FOLDER PATH> <USER_ID>/<MODEL_NAME>

> ```

Reviewer 4: What would be the use of this in the context of mlx-lm use? I'd maybe just say that supported models include the Mistral, Llama-2 and Phi 2 architectures.

Author: But this will change in the future, no?

Author: @pcuenca the CLI was the reason why I didn't want to add push-to-hub since it's not specific to MLX. I addressed this comment.

## Converting and Sharing Models

> You can convert LLMs from Hugging Face Hub as follows: 

Reviewer 4: ```suggestion
You can convert, and optionally quantize, LLMs from the Hugging Face Hub as follows: 
```

```bash
python -m mlx_lm.convert --hf-path mistralai/Mistral-7B-v0.1 -q 
```

If you want to directly push the model after the conversion, you can do it like below. 

```bash
python -m mlx_lm.convert \
    --hf-path mistralai/Mistral-7B-v0.1 \
    -q \
    --upload-repo <USER_ID>/<MODEL_NAME>

> ```

Reviewer 2: this means push to hub is suported in the main table in `/models-libraries` no?

Reviewer 4: Yes, and download from the Hub is supported too.

## Additional Resources

> * [MLX Repository](https://github.com/MaartenGr/BERTopic)

Reviewer: ```suggestion
* [MLX Repository](https://github.com/ml-explore/mlx)
```

Author: sorry I used BERTopic as a template, must've missed this

* [MLX Examples](https://github.com/ml-explore/mlx-examples/tree/main)

> * [MLX-LM](https://github.com/ml-explore/mlx-examples/tree/main/llms/mlx_lm)

Reviewer 4: Also maybe point to the README doc here: https://github.com/ml-explore/mlx-examples/tree/main/llms

Reviewer 3: And maybe also the overall MLX docs:

```
* [MLX Docs](https://ml-explore.github.io/mlx/)
```
